### PR TITLE
Initial commit of N number of objects functionality

### PR DESCRIPTION
**Pull request by callen-sfdo**
Adds fields, new classes, and changes to existing BDI Processing that will allow users to define N number of Additional objects to be created in JSON format (defined by BDI_DynamicSourceGroup).  These dynamic sources will have their processing results stored back into the text field in JSON format rather than being placed in its own field on the DataImport record.  Note that the objects created via this method must children or parents of an existing object mapping and can't currently be child/parents of other dynamically sourced additional objects.

# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by callen-sfdo** [comment]
@melissabarber The needed review from you is on the new Additional_Object_JSON__c field on the DataImport__c object

**Comment by callen-sfdo** [comment]
**lurch: attach W-036291

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001CBBSQA4" target="_blank">W-036291</a>

File: src/objects/DataImport__c.object [hunk 1]
```diff
@@ -258,6 +258,17 @@
         <type>Text</type>
         <unique>false</unique>
     </fields>
+    <fields>
+        <fullName>Additional_Object_JSON__c</fullName>
+        <description>This field allows for additional objects to be specified in a JSON format so that dynamic numbers of records can be created from a single Data Import record.</description>
```
Comment (melissabarber) [inline src/objects/DataImport__c.object]:
```suggestion
        <description>This field allows additional objects to be stored in a JSON format so dynamic numbers of records can be created from a single Data Import record.</description>
```

File: src/objects/DataImport__c.object [hunk 2]
```diff
@@ -258,6 +258,17 @@
...
+        <externalId>false</externalId>
+        <inlineHelpText>This field allows for additional objects to be specified in a JSON format so that dynamic numbers of records can be created from a single Data Import record.</inlineHelpText>
```
Comment (melissabarber) [inline src/objects/DataImport__c.object]:
```suggestion
        <inlineHelpText>This field allows additional objects to be stored in a JSON format so dynamic numbers of records can be created from a single Data Import record.</inlineHelpText>
```

File: src/objects/DataImport__c.object (hunk 1 above)
Comment (melissabarber) [reply src/objects/DataImport__c.object]: @callen-sfdo Mark and I have taken a crack at this. How's it look to you?

File: src/classes/BDI_AdditionalObjectService.cls [hunk 3]
```diff
@@ -123,15 +130,25 @@ public with sharing class BDI_AdditionalObjectService {
                     BDI_ObjectWrapper[] objWrapsForProcessing = new BDI_ObjectWrapper[]{};
                     for (DataImport__c dataImport : dataImportRecords) {
 
+                        BDI_DynamicSourceGroup dsg;
+                        BDI_DynamicSource ds;
+
+                        if (objMapping.dynamicSourceField != null && objMapping.dynamicSourceField.length() > 0) {
+                            dsg = retrieveDynamicSourceDataForDI(dataImport,objMapping.dynamicSourceField);
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply src/classes/BDI_AdditionalObjectService.cls]: Hmm I see how you are using it later without actually setting a DSG ... consider this a maybe-suggestion

File: src/classes/BDI_AdditionalObjectService.cls [hunk 4]
```diff
@@ -469,8 +596,14 @@ public with sharing class BDI_AdditionalObjectService {
         // the creation of this object.
         for (BDI_FieldMapping fieldMapping : objWrap.fieldMappings) {
 
-            if (!importedRecordFieldNames.contains(fieldMapping.Source_Field_API_Name.toLowerCase())
-                && objWrap.dataImport.get(fieldMapping.Source_Field_API_Name) != null) {        
+            if (objWrap.dynamicSource == null) {
+
+                if (!mappingServiceAdvanced.importedRecordFieldNames.contains(fieldMapping.Source_Field_API_Name.toLowerCase())
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService_TEST.cls [hunk 5]
```diff
@@ -1277,4 +1271,541 @@ private class BDI_AdditionalObjectService_TEST {
...
+    static void dynamicParentAdditionalObjectsShouldBeCreated() {
+        Data_Import_Settings__c dis = UTIL_CustomSettingsFacade.getDataImportSettings();
+        dis.Field_Mapping_Method__c = BDI_DataImportService.FM_DATA_IMPORT_FIELD_MAPPING;
+        dis.Default_Data_Import_Field_Mapping_Set__c = 
+            BDI_MappingServiceAdvanced.DEFAULT_DATA_IMPORT_FIELD_MAPPING_SET_NAME;
+        UTIL_CustomSettingsFacade.setDataImportSettings(dis);
+
+        DataImport__c[] diRecords = Database.query(BDI_DataImportService.strSoqlForBatchProcess(null));
+
+        BDI_DataImportService bdiDIS = new BDI_DataImportService(false, BDI_DataImportService.getDefaultMappingService());
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService_TEST.cls]: add line breaks if possible (also line 1295, 1300

File: src/classes/BDI_AdditionalObjectService_TEST.cls [hunk 6]
```diff
@@ -1277,4 +1271,541 @@ private class BDI_AdditionalObjectService_TEST {
...
+                                    FROM DataImport__c]) {
+
+            if (di.Contact1_Lastname__c == 'TestGroupA01') {
+                testDataImportAResult = di;
+            } else if (di.Contact1_Lastname__c == 'TestGroupB01'){
+                testDataImportBResult = di;
+            }
+        }
+
+        BDI_DynamicSourceGroup dsgA =
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService_TEST.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService.cls [hunk 7]
```diff
@@ -312,6 +326,117 @@ public with sharing class BDI_AdditionalObjectService {
...
+    private void updateDIwithDynamicSourceGroups(){
+
+        for (DataImport__c dataImport : dataImportService.listDI) {
+
+            if (dynamicSourceGroupByNameByDataImportId.get(dataImport.Id) != null) {
+                Map<String,BDI_DynamicSourceGroup> dynamicSourceGroupByName = 
+                    dynamicSourceGroupByNameByDataImportId.get(dataImport.Id);
+
+                for (String dynamicSourceFieldName : dynamicSourceGroupByName.keySet()){
+                    dataImport.put(dynamicSourceFieldName,JSON.serialize(dynamicSourceGroupByName.get(dynamicSourceFieldName)));
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService.cls [hunk 8]
```diff
@@ -44,8 +44,13 @@ public with sharing class BDI_AdditionalObjectService {
     private BDI_MappingServiceAdvanced mappingServiceAdvanced;
 
     private Map<Integer,Map<String,BDI_ObjectMapping>> objMappingsByTier = new Map<Integer,Map<String,BDI_ObjectMapping>>();
-    private Map<String,Map<String,BDI_ObjectMapping>> successorObjMappingsByPredecessorName = new Map<String,Map<String,BDI_ObjectMapping>>();
+    private Map<String,Map<String,BDI_ObjectMapping>> successorObjMappingsByPredecessorName = 
+        new Map<String,Map<String,BDI_ObjectMapping>>();
     private Set<String> importedRecordFieldNames = new Set<String>();
+    public Set<String> dynamicSourceFields = new Set<String>{DataImport__c.Additional_Object_JSON__c.getDescribe().getName()};
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService_TEST.cls [hunk 9]
```diff
@@ -1277,4 +1271,541 @@ private class BDI_AdditionalObjectService_TEST {
...
+    * when using a dynamic JSON data source.  Also tests error handling for Dynamic sources.
+    */
+    @isTest static void shouldCreateChildDynamicAdditionalObjects() {
+        Data_Import_Settings__c dis = UTIL_CustomSettingsFacade.getDataImportSettings();
+        dis.Field_Mapping_Method__c = BDI_DataImportService.FM_DATA_IMPORT_FIELD_MAPPING;
+        dis.Default_Data_Import_Field_Mapping_Set__c = 
+            BDI_MappingServiceAdvanced.DEFAULT_DATA_IMPORT_FIELD_MAPPING_SET_NAME;
+        UTIL_CustomSettingsFacade.setDataImportSettings(dis);
+
+        String additionalObjJSONFieldName = UTIL_Namespace.alignSchemaNSWithEnvironment('npsp__Additional_Object_JSON__c');
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService_TEST.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService_TEST.cls [hunk 10]
```diff
@@ -1277,4 +1271,541 @@ private class BDI_AdditionalObjectService_TEST {
...
+        String templateName = 'GAU_Allocation_1';
+        String templateNameOCR = 'Opportunity_Contact_Role_1';
+
+        BDI_DataImportService bdiDIS = new BDI_DataImportService(false, BDI_DataImportService.getDefaultMappingService());
```
Comment (ptewson-sfdo) [inline src/classes/BDI_AdditionalObjectService_TEST.cls]: Add line breaks if possible

File: src/classes/BDI_AdditionalObjectService.cls (hunk 3 above)
[the remark this answers is not in the document]
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService.cls]: I will hold off on this for now.

File: src/classes/BDI_DynamicSource.cls [hunk 11]
```diff
@@ -0,0 +1,12 @@
+public with sharing class BDI_DynamicSource {
+    @AuraEnabled public String objectMappingDevName;
+    @AuraEnabled public String objectMappingTemplateDevName;
+    @AuraEnabled public String fieldMappingSetDevName;
+    @AuraEnabled public Id importedRecordId;
+    @AuraEnabled public String importedRecordStatus;
+    @AuraEnabled public SObject sourceObj;
+
+    public BDI_DynamicSource() {
```
[the remark this answers is not in the document]
Comment (callen-sfdo) [reply src/classes/BDI_DynamicSource.cls]: Done and I also added a class header.

File: src/classes/BDI_DynamicSourceGroup.cls [hunk 12]
```diff
@@ -0,0 +1,8 @@
+public with sharing class BDI_DynamicSourceGroup {
+    @AuraEnabled public Map<String,BDI_DynamicSource> dynamicSourceByObjMappingDevName;
+    @AuraEnabled public String name;
+    
+    public BDI_DynamicSourceGroup() {
```
[the remark this answers is not in the document]
Comment (callen-sfdo) [reply src/classes/BDI_DynamicSourceGroup.cls]: Done and I also added a class header.

File: src/classes/BDI_AdditionalObjectService_TEST.cls (hunk 6 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService_TEST.cls]: Done

File: src/classes/BDI_AdditionalObjectService.cls (hunk 7 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService.cls]: Updated

File: src/classes/BDI_AdditionalObjectService.cls (hunk 8 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService.cls]: Updated

File: src/classes/BDI_AdditionalObjectService_TEST.cls (hunk 9 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService_TEST.cls]: Updated

File: src/classes/BDI_AdditionalObjectService_TEST.cls (hunk 10 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService_TEST.cls]: Updated

File: src/classes/BDI_AdditionalObjectService.cls (hunk 4 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService.cls]: Updated

File: src/classes/BDI_AdditionalObjectService_TEST.cls (hunk 5 above)
Comment (callen-sfdo) [reply src/classes/BDI_AdditionalObjectService_TEST.cls]: Updated

**Pull request merged**